tendermint/src/block/id.rs

```diff
@@ -34,12 +40,92 @@ pub struct Id {
     /// way to propagate a large file over a gossip network.
     ///
     /// <https://github.com/tendermint/tendermint/wiki/Block-Structure#partset>
-    pub parts: Option<parts::Header>,
+    pub parts: Option<PartSetHeader>,
+}
+
+impl DomainType<RawBlockId> for Id {}
+
+impl TryFrom<RawBlockId> for Id {
+    type Error = Error;
+
+    fn try_from(value: RawBlockId) -> Result<Self, Self::Error> {
+        Ok(Self {
+            hash: value.hash.try_into()?,
+            parts: match value.part_set_header {
+                None => None,
+                Some(p) => Some(p.try_into()?),
+            },
+        })
+    }
+}
+
+impl From<Id> for RawBlockId {
+    fn from(value: Id) -> Self {
+        // https://github.com/tendermint/tendermint/blob/1635d1339c73ae6a82e062cd2dc7191b029efa14/types/block.go#L1204
+        // The Go implementation encodes a nil value into an empty struct. We try our best to
+        // anticipate an empty struct by using the default implementation which would otherwise be
+        // invalid.
+        if value == Id::default() {
+            RawBlockId {
+                hash: vec![],
+                part_set_header: Some(RawPartSetHeader {
+                    total: 0,
+                    hash: vec![],
+                }),
+            }
+        } else {
+            RawBlockId {
+                hash: value.hash.into(),
+                part_set_header: value.parts.map(|p| p.into()),
+            }
+        }
+    }
+}
+
+impl TryFrom<RawCanonicalBlockId> for Id {
```

Comment:
We don't use it. There are multiple weak reasons why it's there:
1. Symmetric implementation of all conversions: `Domain::TryFrom(Raw)` and `Raw::From(Domain)`. This sets a good precedent for the developer and standard expectations for all types and conversions. It's a weak reason because if we know that we don't use one conversion, we could simply omit it from the code. I kept it because this is a library and I wasn't sure who uses it for what.
2. The above symmetry is a prerequisite for the domaintype trait. It's a weak argument, because the trait is only available with exactly one raw type, and in this case, it's implemented with the raw BlockId and not the CanonicalBlockId. So, there's no real need for the same symmetric implementation because the DomainType trait doesn't really depend on it.

So it's there for good measure. We can remove it, if unnecessary.